### ISSUE TITLE
This is a leading edge only library, right? Possible have trailing edge?

### ISSUE THREAD

**Issue by Barabba11** [AJMansfield/TriacDimmer#25]
Hi mate, your library seams fantastic, I was looking for something that hardware-directly drives the power stage, leaving CPU for other purposes. I'm going to test it soon :) thanks for sharing!
As designed to drive a Triac I guess it waits for some time before give the pulse to gate, this is leading edge and it works really bad (or almost doesn't work at all for low dims) for dimmable led bulbs, which are my goals, and I suppose the goal in the future for many more others.

I've designed a small cicruit with mosfets to drive trailing edge a limited power lamps, can be around 500W (which is a lot, anyway), I will be happy to share it with you if that is interesting.
By the way I need an inverse logic for it, means the gate should be triggered as soon the zero is crossed, and be maintened (not a pulse only) to desidered dim value.
Being more accurate, I would say my zero cross hardware has about 800uS of detection, I mean the period of time from the first HIGH readed by pin (which trigger interrupt) and the first LOW readed when the wave starts increasing again. This interval is determined by the choice of componets, aimed to reduce hardware and flalse tiggering and noise, not really important.. I just know the zero will start about 400uS after it, nothing is critical here as the voltage is too low to drive anything.

I'm going to read your code, hope to understand as my level of C++ is not that great, hopefully I can myself change something.. 
Thanks for your kind support, I hope it's interesting for you!

**Comment by Barabba11** [comment]
![Dimmer Trailing Gio2](https://github.com/user-attachments/assets/89877733-945a-4604-9041-86af9ab954df)

This is the circuit I've designed, it's efficent, works with low temperatures and consuming 0,147W on st-by, mosfets have only 0,57ohm resistance and support up to 10A (but I recoment use till 3A, in this case you can use a small heat dissipator and you can put the 2 mosfets on the same radiator (the good thing here is that 2SK3569 has insulated body) to reduce the circuit dimensions, two extra diodes are used to reduce the forward voltage of the internal mosfet's diode, lowering mosfet's dissipation and working temperature, they can be eventually avoided if the load is <1A
The circuit is driven by 2 optocouplers, one is used to activate mosfets, the second to discharge the gates quickly, leaving mosfets operate on saturation only.
The microcontroller used should be loaded enough to assure optocouples create the most available saturation of transistors, to fast charge/discharge gates. For 5V microcontroller a resistor of 390 ohm assures about 13mA for light the opto's leds, on low and high; for 3,3v controllers use 220/270 ohm. The circuits activates the load with LOW logic and turn mosfets off with HIGH level, this is preferable according the fact microcontrollers boot usually with pins on input mode and high pullup internal resistor. Only a forced LOW level can turn on mosfets.
Snubber is optional, by the way it's better to use anycase to preserve life.
I've added the zero crossing detector, I have to work on it more, probably I'll insert a capacitor to avoid noise, maybe it's not necessary an EMI filter. I've to focus to optimal values for resistors to reduce dissipation without enlarging too much the zero crossing LOW time detect.
The zero crossing takes about 0,5W dissipation, I guess it's possible to combine it with the power needed to turn mosfets but the result won't change too much

**Comment by Barabba11** [comment]
![Dimmer Trailing Gio2](https://github.com/user-attachments/assets/aae7d63b-ed54-42e9-8d2b-0d1a695ca554)

**Comment by AJMansfield** [comment]
Glad to hear my little library is something of interest to you! That's a pretty cool circuit, and a neat way to ensure your power mosfets only need flyback protection in one direction, with the efficiency savings of not having to constantly waste a bit of apparent and real power on a snubber network.

First off, as a quick hack to change the output from a short pulse to a continuous level change, you can set the library's `pulse_length` parameter to 65535.
The library's logic is defensive enough that it should still handle that case correctly and end the pulse in the zero-crossing interrupt (and appropriately cancel the pending pulse-off event that the pulse-on interrupt would've queued in hardware).

I'll note that the timing performance is degraded very slightly in that mode -- in that, the pulse only switches off once the ISR runs the associated assembly instruction, so that switch-off isn't protected against interrupt latency the way the normally pulse-on and pulse-off are when they're done entirely from hardware.

If you want to invert the polarity of the pulse in software, that's also possible, if you change the values written to the `COM1x0` and `COM1x1` control bits for each channel to change 'set' operations into 'clear' operations and vice versa. (I'd highly recommend familiarizing yourself with Chapter 15 of the [Atmega328p datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-7810-Automotive-Microcontrollers-ATmega328P_Datasheet.pdf) first, if you're going to do this.)

As a potentially simpler way of inverting the polarity, though, why not just swap the two optocouplers around to change which state drives which side?

![image](https://github.com/user-attachments/assets/d5386f9c-ccde-41b6-904c-4372ba0586d5)

(I can only assume, btw, that the top diode oriented with cathode to VCC is just a schematic error -- neither state of the ardunio's pin would ever make that diode emit as drawn.)

One other thought, is just to make sure you're considering shoot-through. I've not done the math to check whether or to what degree shoot-through could cause a problem for your circuit, but with those sorts of bridge circuits, it's usually worth making sure you do indeed have enough sufficient clearance time.

**Comment by AJMansfield** [comment]
Regarding the shoot-through issue: (posting this slightly better explanation separately in case you'd have missed an edit)

Specifically, as the control signal rises and falls, it has to pass through intermediate voltages -- the pin isn't really turning _on_ the optocouplers by providing a current path, as much as it's inhibiting them by keeping the voltage difference from that center node to VCC or GND below the optocoupler's forward voltage. During the rise/fall time, it's probably inhibiting neither diode; and even if it's fast enough to keep that intermediate region sufficiently short, it's still going to let the currently-disabled diode start emitting before the other optocoupler's gate has stopped conducting.

With both optocouplers conducting at the same time for short periods, that'll at best slow down the rise and fall time on the mosfet gates, possibly leaving them in their linear region for longer than expected.

Again, I've not done the math to see if this is a problem for your specific component choices, but if it is, there's two things that can be done about it. First is to add some extra series diodes to each optocoupler diode, so the combined forward voltage of both strings exceeds the total supply voltage. Once you've done that, then if the switch-off time for the optocoupler gates still exceeds the clearance time, you can then make extra clearance time by slowing down the Atmega's output edge with an extra few pF of capacitance.

**Comment by Barabba11** [comment]
Hi mate! I'm really glad you liked this project and you wrote me your considerations :) Since you are here and I don't want to loose the chance to reply you quickly, I would for now write what I think I need to clear deeper:
- I assume the pulse parameter at max value isn't influence any timings, just ignore to output as pulse, but I would you keep in consideration that they only time I'm having the line "active" (low in my case, mosfets driven ON) is when I want 100% of brightness. My circuit has the opposite logic than yours, the ON starts after zero and keeps for the needed % of brightness, I want be sure you have clear idea how trainilg edge dimmers work, sorry if for you is all already well known :) I post a picture, maybe useful to who is reading here too.
So, when you confim that what you wrote is according trailing edge I'll read deeper your instructions and have a try :)
The diodes are designed right, a LOW output turns on the opto above, leaving 15V trigger mosfets on, an HIGH output turns on the opto behind, discharging quickly the gates of mosfets and interrupting the power to load.
The reason why I want an HIGH as power off is that controllers usually boot with pins set as input with pullup resistor. Choosing LOW is much safer way

![trailing](https://github.com/user-attachments/assets/7a9e10ad-e6a0-438c-9d2e-6b2d8cf9bf6e)

**Comment by Barabba11** [comment]
About output and opto diodes, your' right, I share it with you, my mistake on design the circuit today, as when I did one year ago.. I can't find more the paper )) as you see in my scheme I put 2 resistors to drive separately the opto, and avoid such problems. Sorry for mistake, thanks to notify that, I correct and upload again the scheme

The circuit perfectly works and the mosfet are keeping cold. The biggest problem of mosfets here is not the speed of charging/discharging capacity and stay in active zone, neither their ON resistance, but it's the internal diode protection, which has a vf of 1,7v (ok at max 10A but still pretty high), means  at 3A about 4,5W dissipation. I'm using them to let the current flow on the inhibited mosfet, but hey are concepted only to kill the negative inverted spices of a coil, for few ms, and there the vf isn't important. Here it is, this is why I put an extra diode outside with a lower vf.

**Comment by Barabba11** [comment]
Reading again your comment about the library work I suppose there is needed to change how the timers operate, considering how the circuit is designed, briefly we can consider:

the zero cross disables both diodes of the opto PC814, leaving the pullup lead the interrupt pin to HIGH, and generate the interrupt. The zero is not crossed yet, turning ON now the mosfet isn't corrected but not dangerous as the low voltage won't turn on anything connected on load output.
Having at setup stage of sketch a simple measure to read time from HIGH to LOW and get an average value could give with great chances to guess the exact moment when the zero will cross. In my case I have an average of 840uS, which is not low but ok, ininfluent.
Knowing that the zero (in my case) will cross 400uS after interrupt, and if the brightness is set >0%, after 400uS the interrupt it will be time to load the counter and get a LOW value on output (that turns one mosfets), till the timer's counter decrease to value set and this will reflect HIGH to output turning mosfets off.

Now I understand it's impossible to manage the first 400uS easly, we may need a second interrupt for that, generated by timer.
Maybe, as I told, better leave that and keep thigs easy, turn on mosfets even if the zero is not reached, just add 400uS before calculate brightness.

More important is:
- 0% of brightness should keep output high and not load timer, for save energy.
- The bulbs should turned on smoothly, and brightness changed smootly too, this is to avoid high peaks currents on mosfets caused by many lamps bulb's capacitors connected in parallel, probably it has more sense to manage that into the loop than the library

I've sent today the circuit to Arduino and Adafruit, I hope they can consider to develope a shield for it, since there is nothing trailng edge available on market :)

**Comment by AJMansfield** [comment]
This library is actually meant specifically for leading-edge dimming, not trailing edge -- as that's the only dimming mode possible with a triac.

Triacs are self-latching bidirectional semiconductor switches -- that is, it only takes a short trigger pulse on the gate pin of a triac to switch it from blocking to conducting, and once triggered it remains conducting region until the current through it drops to zero, where it then switches back off automatically.

All that this library does, when controlling a triac, is precisely measure the time interval between the previous two zero crossings, and then schedule the PWM timer to deliver a short voltage pulse to the triac circuit, to trigger it to push the triac into the conducting state for the rest of that half-wave -- it's not even involved in switching the triac off at the next zero.

![EWPE8nX](https://github.com/user-attachments/assets/ae96e3e0-c819-455f-996f-98b669980463)

The `pulse_length` parameter exists primarily to control how long these pulses are. Too short of a pulse may not manage to trigger it, while in theory too long of a pulse can over-saturate a triac's trigger.

But, if you instead set that pulse length to be extra long, you can effectively generate an in-phase square wave instead.

The comment about timing is just to note that those falling edges of that square wave won't be exactly aligned with the zero-crossing signal, and isn't guaranteed to even have exactly the same offset from it every time, due to interrupt latency jitter.
Though, odds are, a few dozen microseconds of jitter are probably irrelevant anyway if you're just controlling some lights.

**Comment by Barabba11** [comment]
Yes, moreover, the zero crossing interrupt is triggered before it cross the zero, so no worries, there is plenty of time to exacute the interrupt routine.
As I see, your library generates a LOW level after the zero cross, that's ok for me, no need to manage COM1x0 and COM1x1?
I just need to invert bightness value (my 100% is your 0%)

**Comment by AJMansfield** [comment]
Regarding shoot-through, here's the math:

Imagine for a moment that you put the arduino's pin into hi-Z as an input, effectively taking it and the current-limiting resistor out of circuit. 
During the transition between low and high on the arduino's output, the pin will pass through whatever voltage that node would have in the high-Z condition, so it's a valid simplification here to just imagine that moment as if the arduino and resistor didn't exist.

A PC123 optocoupler has an input diode with a forward voltage of 1.2 volts. Two of them in series, gives 2.4 volts, and since that's less than 3.3, both diodes are going to be conducting.
Each one will drop half of the 3.3, i.e. 1.65, and if you look at the datasheet voltage vs current charts, you can pick out how much forward current each diode will conduct at 1.65 volts:

![image](https://github.com/user-attachments/assets/a6281a32-909d-4bca-a599-4d6387db74d7)

Depending on temperature, they're both conducting 80 -150 mA during that condition -- not just 'on', but actually above the 50 mA maximum continuous current rating.

Based on the ratings of the Atmega328p (40mA max pin current, 200ns max free rise and fall time) and the PC123 (250 pF max input capacitance), the RC-time constant here that controls how long that middle node spends near this mid-voltage, should be on the order of 250ns. Figuring out the exact threshold voltages to calculate the exact dwell time for the double-conducting condition would be a lot more math, but fortunately we can disregard that because the rise and fall time specs on the PCA123s themselves are already worse than that.

There's no minimum rise or fall time, only a typical 4us rise and 3us fall and a maximum 18us for both. It would, be in spec for the slower of the two PCA123's to still be conducting for another 18us while the faster of the two turns on instantly, giving a worst-case shoot-through of 18 us. Using the typical values, though, the typical shoot-through should be around -1 us. (Basically: optocouplers are semiconductor devices that have their own non-ideal saturation behaviors.)

So as long as the arduino's edge is fast enough, there should be a _bit_ of clearance time in the typical case. I'd just caution against being optimistic about that rise/fall asymmetry, if you mean to productize this.

Also, that maximum current spec on the diodes is something you really ought to pay attention to. Outside that hi-Z equivalent condition, the current through the diode that's on is going to be _higher_ than the hi-Z condition current -- so you're violating the max current spec regardless of how large of a resistor you fit between the arduino's pin and that center node. You need _two_ separate current-limiting resistors here.

**Comment by AJMansfield** [comment]
> Yes, moreover, the zero crossing interrupt is triggered before it cross the zero, so no worries, there is plenty of time to exacute the interrupt routine. As I see, your library generates a LOW level after the zero cross, that's ok for me, no need to manage COM1x0 and COM1x1? I just need to invert bightness value (my 100% is your 0%)

Or, swap the anode and cathode of both outbound optocoupler diodes, and swap the VCC and GND at either end; that'll change which optocoupler is being activated during each polarity, rather than needing to change the polarity of the signal itself.

**Comment by Barabba11** [comment]
Yes mate :) As I wrote you were right, the protoype I realized has 2 resistors, it should be like that, my scheme was wrong and I've already corrected it :)
talking about 2 resistors, the z condition means very low currents, that eventually can reflect on drive ground the gates, this is not dangerous, at the same time the upper diode won't let pass vcc on second diode, so its an open circuit.
About transient, a totem pole output will give vcc or ground (talking ideally, but here doesn't matter the real values), so as the z condition is not dangerous we can say that led up works only when totem pole energize it with ground and led down when energizes with vcc. No collisions, no active area, in my opinion perfect solution :)

About inverting, much more easier is to change software )) 
As I wrote, I'll keep LOW level to activate Mosfets, this is safer.

**Comment by AJMansfield** [comment]
>     * 0% of brightness should keep output high and not load timer, for save energy.


Note, the library does actually already do those 0% and 100% special cases. Anything above the `on_thresh` setup parameter value will just disable the timer pulsing and set the output high, and anything below the `off_thresh` value disables timer pulsing an sets the output low.

>     * The bulbs should turned on smoothly, and brightness changed smootly too, this is to avoid high peaks currents on mosfets caused by many lamps bulb's capacitors connected in parallel, probably it has more sense to manage that into the loop than the library

To be honest I'm a bit skeptical of the utility of that, but I'm sure you can figure out how to implement the behavior you want with the available APIs.

Note, the `TriacDimmer::detail::` namespace exists precisely to provide low-level access to the underlying timer ISR code. If you have specific needs for how the dimming should work, I'd recommend writing your own wrapper that calls `TriacDimmer::begin` just to do initial register setup, and then writes to the control values in `TriacDimmer::detail::` directly to enable/disable the channels and set the delays.

These detail-namespace variables can be directly set, and get read by the ISRs as their means of learning what the non-ISR-land user logic wants to have happen: `ch_A_en`, `ch_A_up`, `ch_A_dn`, `ch_B_en`, `ch_B_up`, `ch_B_dn`

And the `period` detail variable is where the ISRs write the measured length of mains frequency in timer-count units, to base the values written to the `*_up` and `*_dn` variables.
